xrdp/xrdp_listen.c

```diff
@@ -771,16 +771,33 @@ xrdp_listen_process_startup_params(struct xrdp_listen *self)
 static int
 xrdp_listen_fork(struct xrdp_listen *self, struct trans *server_trans)
 {
+    char server_trans_fd_str[32];
+    char executable_path[4096];
+    struct list *child_arguments;
+
     int pid;
     int index;
     struct xrdp_process *process;
     struct trans *ltrans;
 
-    pid = g_fork();
+    g_snprintf(server_trans_fd_str, 32, "%d", (int) server_trans->sck);
+
+    child_arguments = list_create();
+    list_add_item(child_arguments, (intptr_t) g_strdup("--child-process"));
+    list_add_item(child_arguments, (intptr_t) g_strdup("--child-fd"));
+    list_add_item(child_arguments, (intptr_t) g_strdup(server_trans_fd_str));
+
+    g_get_executable_path(executable_path, 4096);
+
+    pid = g_fork_execvp(executable_path, (char **) child_arguments->items);
```

Comment:
Memory leaks in the above as you have a mix of static and dynamic allocations. It doesn't matter for the child but the parent will accumulate the losses over time.

If you set `child_arguments->auto_free = 1`, memory is deallocated when the list is deleted. That means everything has to be allocated, but it's easy to do.

Have a look at neutrinolabs/xrdp#2576 which is merged recently. That lets you do the lot with this, or something similar. As a benefit, memory allocations are checked:-

```c
    child_arguments = list_create();
    if (child_arguments != NULL)
    {
        child_arguments->auto_free = 1;
        if (!list_add_strdup_multi(child_arguments,
            "--child-process",
           "--child-fd", server_trans_fd_str,
          NULL))
        {
            list_delete(child_arguments);
            child_arguments = NULL;
        }
    }

    if (!child_arguments)
    {
        // Error code
    }
    else
    {
        // Use list
       list_delete(child_arguments);
    }
```